### PR TITLE
SWATINIT QC tool

### PR DESCRIPTION
**Pull request by berland**
New tool for QC of SWATINIT (water initialization in Eclipse).

Aims:
* Produce a human readable report
* Produce a CSV with data pr. cell  that can easily be used by other visualization tools to present the same results, both in 2D and 3D.

Example text output:
```
$ check_swatinit ~/projects/drogon/20.4.0/eclipse/model/DROGON_HIST.DATA
VOLUME                     3203.1103 Mrm3
PORV                        504.6057 Mrm3
SWATINIT_WVOL               506.5309 Mrm3           HC:   66.571 Mrm3
+ EQUIL_INIT                  0.0000 Mrm3   0.00 %         0.00 %
+ HC_BELOW_FWL                0.6500 Mrm3   0.13 %         -0.98 %
+ PPCWMAX                     0.0000 Mrm3   0.00 %         0.00 %
+ SWATINIT_1                  0.0000 Mrm3   0.00 %         -0.00 %
+ SWL_TRUNC                   1.2752 Mrm3   0.25 %         -1.92 %
= SWAT_WVOL                 506.5309 Mrm3   0.38 %         -2.89 %

Maximal values:
---------------
        PCOW_MAX
SATNUM          
1        9.95715
2        9.95715
3        9.95715
4        9.95715
7        9.95715
8        9.95715
9        9.95715
10       9.95715
11       9.95715
12       9.95715
                     PPCW  PC_SCALING
EQLNUM SATNUM                        
1      1       115.018127   11.551310
       2        74.829216    7.515124
       3       150.559143   15.120707
       7        28.181242    2.830252
       8        63.305759    6.357819
       9              NaN         NaN
       10       26.814203    2.692960
       11       18.821041    1.890204
       12       13.947854    1.400788
2      1       266.361938   26.750821
       2        67.441399    6.773163
       3        87.531380    8.790807
       7        33.537632    3.368196
       8        56.655766    5.689958
<snip>
```

image

image

### CODE REVIEW

File: docs/scripts/check_swatinit.rst [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  significant volumes, revise the modelling.
+
+``SWATINIT_1``
+  When SWATINIT is 1 above the contact, Eclipse will ignore SWATINIT in the cell
+  and not touch the capillary pressure function. This will typically result in extra
+  hydrocarbons added to the model for a normal capillary pressure function. This
+  could be ok as long as the porosities and/or permeabilities of these cells are
+  small. If it is not, you should look into if there are upscaling issues for
+  this cell. In situations with nonzero item #9 in EQUIL, this can also occur
+  below the contact.
```
Comment (vkip) [inline docs/scripts/check_swatinit.rst:76]: Could mention that having Pc(Sw=1)=epsilon>0 allows Eclipse to match SWATINIT=1 also above the contact.

File: src/subscript/check_swatinit/check_swatinit.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+        and "SWATINIT" not in deck
+    ):
+        logger.warning(
+            "INIT-file/deck does not have SWATINIT, this tool has limited use."
+        )
+
+    if "RPTRST" not in deck:
+        logger.warning(
+            "RPTRST not found in DATA-file, UNRST file is expected to be missing"
+        )
```
Comment (vkip) [inline src/subscript/check_swatinit/check_swatinit.py:123]: Probably need to check that RPTRST is present in the SOLUTION section to ensure restart output at time 0?

File: src/subscript/check_swatinit/check_swatinit.py [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+        ] = __PC_SCALED__
+
+    # PC is scaled (including "scaling" using PC_SCALING=1), but
+    # SWAT!=SWATINIT, this must be due to EQUIL item 9 being nonzero.
+    if "OIP_INIT" in qc_frame and "PC_SCALING" in qc_frame:
+        qc_col[
+            (~np.isclose(qc_frame["OIP_INIT"], 0))
+            & (~np.isclose(qc_frame["SWAT"], qc_frame["SWATINIT"], atol=1e-6))
+            & (~pd.isnull(qc_frame["PC_SCALING"]))
+        ] = __EQUIL__
```
Comment (vkip) [inline src/subscript/check_swatinit/check_swatinit.py:344]: __EQUIL__ might not be the most descriptive label... Perhaps '__FINEGRID_EQUIL__' , '__FINE_EQUIL__' or some better suggestion?

File: src/subscript/check_swatinit/pillarmodel.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
+import textwrap
```
Comment (vkip) [inline src/subscript/check_swatinit/pillarmodel.py:1]: Move to tests, as it is not used by the script (as far as I could see)?

File: src/subscript/check_swatinit/check_swatinit.py (hunk 2 above)
Comment (berland) [reply src/subscript/check_swatinit/check_swatinit.py:123]: Do you know how to ask a `DeckKeyword` for which section it belongs to?

File: docs/scripts/check_swatinit.rst (hunk 1 above)
Comment (berland) [reply docs/scripts/check_swatinit.rst:76]: Yes, I think this falls into the PC_SCALED category, with "scaling" to 1.

File: src/subscript/check_swatinit/check_swatinit.py (hunk 3 above)
Comment (berland) [reply src/subscript/check_swatinit/check_swatinit.py:344]: The label the user will see is currently `EQUIL_INIT`. But `FINE_EQUIL` possibly sounds better.

File: src/subscript/check_swatinit/pillarmodel.py (hunk 4 above)
Comment (berland) [reply src/subscript/check_swatinit/pillarmodel.py:1]: Correct that `pillarmodel.py` is only used by tests for now. Having it deployed alongside check_swatinit allows usage elsewhere, while if it lies in tests, it cannot be used by others. Not sure about this.

File: src/subscript/check_swatinit/check_swatinit.py (hunk 2 above)
Comment (vkip) [reply src/subscript/check_swatinit/check_swatinit.py:123]: No.. I guess just checking if report step 0 is present is easier.

File: src/subscript/check_swatinit/pillarmodel.py (hunk 4 above)
Comment (vkip) [reply src/subscript/check_swatinit/pillarmodel.py:1]: See the point - perhaps there should be common place for cross-script tools?

File: src/subscript/check_swatinit/check_swatinit.py (hunk 2 above)
Comment (berland) [reply src/subscript/check_swatinit/check_swatinit.py:123]: There is a `test_no_rptrst()` that checks that you get a `SystemExit` mentioning `UNRST` already if there is no data at report step 0. The code above is only meant as a hint.

**Review by vkip — APPROVED** [review]
Great work - comments are nitpicking..

File: src/subscript/check_swatinit/pillarmodel.py (hunk 4 above)
Comment (berland) [reply src/subscript/check_swatinit/pillarmodel.py:1]: Kept as is.

**Pull request merged**